File: src/SkyBlock/tiles/AutoSellerTile.php
<?php

namespace SkyBlock\tiles;

use pocketmine\block\tile\Chest;
use pocketmine\block\tile\Nameable;
use pocketmine\block\tile\NameableTrait;
use pocketmine\block\tile\Spawnable;
use pocketmine\item\Item;
use pocketmine\math\Facing;
use pocketmine\nbt\tag\CompoundTag;
use pocketmine\player\Player;
use pocketmine\Server;
use SkyBlock\block\AutoSeller;
use SkyBlock\Main;

class AutoSellerTile extends Spawnable implements Nameable {

    use NameableTrait;

    public const TAG_LEVEL = "Level";
    public const TAG_TYPE_XP = 1;
    public const TAG_TYPE_MONEY = 0;
    public const MAX_LEVEL = 8;
    public const TAG_TYPE = "Type";
    public int $delay = 200, $level1 = 0, $type = 0;

    public static function getTypeName(int $type) : string {
        return ($type === self::TAG_TYPE_MONEY) ? "money" : "xp";
    }

    public function onUpdate() : bool {
        if ($this->level1 === 0) $this->close();
        if ($this->closed) return false;
        $block = $this->getBlock();
        if (!$block instanceof AutoSeller) return false;
        $inst = Main::getInstance();
        if (($island = $inst->getIslandManager()->getOnlineIslandByWorld($this->getPosition()->getWorld()->getDisplayName())) === null) return false;
        if (!isset($inst->autoseller[$this->getPosition()->getWorld()->getDisplayName()])) {
            if ($this->delay <= 0) {
                $this->delay = $this->getDelayByLevel($this->level1);
                $down = $this->getPosition()->getWorld()->getTile($this->getBlock()->getSide(Facing::DOWN)->getPosition());
                if ($down instanceof Chest) {
                    $inv = $down->getInventory();
                    foreach ($inv->getContents() as $item) {
                        if ($item instanceof Item) {
                            //                                var_dump(Main::getInstance()->getEvFunctions()->getSellMoneyData($item->getVanillaName()));
                            if (Main::getInstance()->getEvFunctions()->getSellMoneyData($item->getVanillaName()) !== 0) {
                                $receiver = $island->getReceiver();
                                $player = Server::getInstance()->getPlayerExact($receiver);
                                if (!$player instanceof Player) {
                                    if (($player = $island->getRandomOnlineCoOwner()) === null) return true;
                                    $receiver = $player->getName();
                                }
                                if (($user = $inst->getUserManager()->getOnlineUser($receiver)) !== null && $island->isAnOwner($receiver)) {
                                    if ($this->type === self::TAG_TYPE_MONEY)
                                        $inst->getEvFunctions()->sellItem($user, $item->setCount($this->level1), "money");
                                    else if ($this->type === self::TAG_TYPE_XP)
                                        $inst->getEvFunctions()->sellItem($user, $item->setCount($this->level1), "xp");
                                    $inv->removeItem($item);
                                    return true;
                                }
                            }
                        }
                    }
                }
            } else {
                --$this->delay;
            }
        }
        return true;
    }

    public function getDelayByLevel(int $level) : int {
        if ($level == 1) return 15 * 20; // 15 seconds
        elseif ($level == 2) return 10 * 20;// 10 seconds
        elseif ($level == 3) return 8 * 20;// 8 seconds
        elseif ($level == 4) return 6 * 20;// 6 seconds
        elseif ($level == 5) return 4 * 20;// 4 seconds
        elseif ($level == 6) return 2 * 20;// 2 seconds
        elseif ($level == 7) return 30; // 1.5 second
        elseif ($level == 8) return 20; // 1 a second
        else return 200;
    }

    public function getName() : string {
        return "AutoSeller";
    }

    public function getDefaultName() : string {
        return "AutoSeller";
    }

    public function readSaveData(CompoundTag $nbt) : void {
        if ($nbt->getInt(self::TAG_LEVEL, null) === null) $this->close();

        $this->level1 = $nbt->getInt(self::TAG_LEVEL, $this->level1);
        $this->type = $nbt->getInt(self::TAG_TYPE, $this->type);

        $this->delay = $this->getDelayByLevel($this->level1);
    }

    public function getDelayInSeconds() : int {
        return (int) ($this->getDelayByLevel($this->level1) / 20);
    }

    public function setDelay(int $delay) : void {
        $this->delay = $delay;
    }

    protected function writeSaveData(CompoundTag $nbt) : void {
        $this->applyBaseNBT($nbt);
    }

    private function applyBaseNBT(CompoundTag $nbt) : void {
        $nbt->setInt(self::TAG_LEVEL, $this->level1);
        $nbt->setInt(self::TAG_TYPE, $this->type);
    }

    protected function addAdditionalSpawnData(CompoundTag $nbt) : void {
        $this->applyBaseNBT($nbt);
    }


}

File: SkyblockCore/src/main/java/dev/skyblock/tiles/AutoSellerTile.java
package dev.skyblock.tiles;

import dev.skyblock.SkyblockCore;
import dev.skyblock.island.Island;
import dev.skyblock.user.User;
import dev.skyblock.util.SellPrices;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.block.Block;
import org.bukkit.block.Chest;
import org.bukkit.entity.Player;
import org.bukkit.inventory.Inventory;
import org.bukkit.inventory.ItemStack;

public class AutoSellerTile extends BaseTile {
    public static final int MAX_LEVEL = 8;
    public static final int TYPE_MONEY = 0;
    public static final int TYPE_XP = 1;

    private int tickCounter = 0;
    private final TileData data;

    public AutoSellerTile(Location location, TileData data) {
        super(location);
        this.data = data;
        this.tickCounter = getDelayTicks(data.getLevel());
    }

    @Override
    public void onUpdate() {
        if (tickCounter > 0) {
            tickCounter--;
            return;
        }
        tickCounter = getDelayTicks(data.getLevel());

        SkyblockCore plugin = SkyblockCore.getInstance();
        String worldName = location.getWorld().getName();

        Island island = plugin.getIslandManager().getOnlineIslandByWorld(worldName).orElse(null);
        if (island == null) return;

        // Get chest directly below this block
        Block below = location.getBlock().getRelative(0, -1, 0);
        if (!(below.getState() instanceof Chest chestState)) return;

        Inventory inv = chestState.getInventory();

        // Find receiver — island owner if online, otherwise a co-owner
        String receiverName = island.getReceiver();
        Player receiver = plugin.getServer().getPlayerExact(receiverName);
        if (receiver == null) {
            receiver = island.getRandomOnlineCoOwner();
            if (receiver == null) return;
            receiverName = receiver.getName();
        }

        User user = plugin.getUserManager().getOnlineUser(receiver.getUniqueId());
        if (user == null || !island.isAnOwner(receiverName)) return;

        // Sell each item in chest
        for (ItemStack item : inv.getContents()) {
            if (item == null || item.getType() == Material.AIR) continue;
            double price = SellPrices.getPrice(item.getType());
            if (price <= 0) continue;

            // Amount sold per cycle = item count * level multiplier
            int sellCount = Math.min(item.getAmount(), data.getLevel());
            double earnings = price * sellCount;

            if (data.getType() == TYPE_MONEY) {
                user.addMoney(earnings);
            } else {
                // XP type — add xp to user
                user.addXp((int) earnings);
            }

            String type = data.getType() == TYPE_MONEY ? "$" : "XP";
            receiver.sendActionBar(net.kyori.adventure.text.Component.text(
                "§l§6>> §eAuto Sold for §6" + dev.skyblock.util.Format.formatMoney(earnings) + type + " §6<<"
            ));

            // Remove sold items
            item.setAmount(item.getAmount() - sellCount);
            if (item.getAmount() <= 0) {
                inv.remove(item);
            }

            // Only sell one item type per cycle to match original behaviour
            break;
        }
    }

    // Delay in ticks per level — matches PHP getDelayByLevel exactly
    public static int getDelayTicks(int level) {
        return switch (level) {
            case 1 -> 15 * 20;
            case 2 -> 10 * 20;
            case 3 -> 8 * 20;
            case 4 -> 6 * 20;
            case 5 -> 4 * 20;
            case 6 -> 2 * 20;
            case 7 -> 30;
            case 8 -> 20;
            default -> 200;
        };
    }

    public TileData getData() { return data; }
}
